src/SharpLearning.Optimization/BayesianOptimizer.cs

```diff
@@ -234,68 +262,69 @@ public OptimizerResult[] Optimize(Func<double[], OptimizerResult> functionToMini
             for (int iteration = 0; iteration < m_iterations; iteration++)
             {
                 // fit model
-                var observations = parameterSets.ToF64Matrix();
+                var observations = parameterSets.ToList().ToF64Matrix();
                 var targets = parameterSetScores.ToArray();
                 var model = m_learner.Learn(observations, targets);
 
                 var bestScore = parameterSetScores.Min();
                 var candidates = FindNextCandidates(model, bestScore);
-                
+
                 var first = true;
 
-                foreach (var candidate in candidates)
+                Parallel.ForEach(candidates, m_parallelOptions, candidate =>
                 {
                     var parameterSet = candidate.ParameterSet;
 
-                    if (Equals(lastSet, parameterSet) && !first)
+                    // skip evaluation if parameters have not changed unless explicitly allowed
+                    if (m_allowMultipleEvaluations || first || !Contains(previousSets, parameterSet))
                     {
-                        // skip evaluation if parameters have not changed.
-                        continue;
-                    }
+                        first = false;
 
-                    if (Equals(bestParameterSet, parameterSet))
-                    {
-                        // if the beset parameter set is sampled again.
-                        // Add a new random parameter set.
-                        parameterSet = RandomSearchOptimizer
-                            .SampleParameterSet(m_parameters, m_sampler);
-                    }
+                        if (!m_allowMultipleEvaluations && Equals(bestParameterSet, parameterSet))
+                        {
+                            // if the best parameter set is sampled again.
+                            // Add a new random parameter set.
+                            parameterSet = RandomSearchOptimizer
+                                .SampleParameterSet(m_parameters, m_sampler);
+                        }
 
-                    var result = functionToMinimize(parameterSet);
-                    iterations++;
+                        previousSets.Add(parameterSet);
 
-                    if (!double.IsNaN(result.Error))
-                    {
-                        // update best
-                        if (result.Error < bestParameterSetScore)
+                        var result = functionToMinimize(parameterSet);
+                        iterations++;
+
+                        if (!double.IsNaN(result.Error))
                         {
-                            bestParameterSetScore = result.Error;
-                            bestParameterSet = result.ParameterSet;
-                            //System.Diagnostics.Trace.WriteLine(iterations + ";" + result.Error);
+
+                            // update best
+                            if (result.Error < bestParameterSetScore)
```

Comment:
@jameschch Same issue here, we need a lock statement for the `if (result.Error < bestParameterSetScore)` check.

As I see it, they only thing we need the `bestParameterSetScore` and `bestParameterSet` for is to check if the model keeps sampling the same parameters over and over again. It might be better to refactor this all together to avoid these two variables, and simply keep a single  `ConcurrentBag<(double[], double>)` and extract the current best solution from this when needed. I think this would reduce the complexity of the code.

Reply:
Have refactored as suggested but needed to use a BlockingCollection. I did try with ConcurrentBag, but this led to strange behaviour.